### PR TITLE
Shopping cart

### PR DESCRIPTION
**Pull request by LHJE**
[Reference story or milestone being closed]

**Contributors**
@LHJE, @leahriffell, @GDemps, @priyapower 
  
**What was changed:**
In cart: 
- I can see a button or link to increment/decrement the count of items I want to purchase
- I cannot increment the count beyond the item's inventory size
- If I decrement the count to 0 the item is immediately removed from my cart
- I see information telling me I must register or log in to finish the checkout process
- The word "register" is a link to the registration page
- The words "log in" is a link to the login page
- I see a button or link indicating that I can check out, and I click the button or link to check out and fill out order info and click create order, an order is created in the system, which has a status of "pending", that order is associated with my user, I am taken to my orders page ("/profile/orders"), I see a flash message telling me my order was created, I see my new order listed on my profile orders page, and my cart is now empty

**Any tech notes (if applicable)**
- Check to see if this typo is still in checkout_spec.rb: `it "does not prompt useres to login before checking out" do`

- Check to see if this is still in index_spec.rb: it "can show 5 most and least popular items and quantity bought" do

### CODE REVIEW

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -9,24 +9,28 @@ def show
   end
 
   def create
-    order = Order.create(order_params)
+    order = current_user.orders.new(order_params)
     if order.save
       cart.items.each do |item,quantity|
-        order.item_orders.create({
+        order.item_orders.create!({
```
Comment (LHJE) [inline app/controllers/orders_controller.rb:15]: Let's take out the bang here!

File: app/controllers/orders_controller.rb [hunk 2]
```diff
@@ -9,24 +9,28 @@ def show
...
           item: item,
           quantity: quantity,
           price: item.price
           })
       end
       session.delete(:cart)
-      redirect_to "/orders/#{order.id}"
+      if current_user.role == "admin"
```
Comment (LHJE) [inline app/controllers/orders_controller.rb:22]: 23 and 24 are ready to be deleted!

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -6,12 +6,14 @@ def new
 
   def create
     @user = User.new(user_params)
-    if @user.save
+    if User.email_exists?(@user.email)
+      @user.email = nil
+      # user_fails_create_validation # This is failing BECAUSE, @user.errors doesn't show ANY errors - we wonder if this is because we had to update a user to "save" without unique email to prompt errors; our current solution is to handroll this flash[:error] message :)
+      flash[:error] = "Email has already been taken"
+      render :new
```
Comment (LHJE) [inline app/controllers/users_controller.rb:13]: There must be a way to get `user_fails_create_validation` to work here again

File: app/models/order.rb [hunk 4]
```diff
@@ -1,6 +1,8 @@
 class Order <ApplicationRecord
   validates_presence_of :name, :address, :city, :state, :zip
+  validates_inclusion_of :status?, :in => [true, false]
```
Comment (LHJE) [inline app/models/order.rb:3]: We still have questions about whether or not we need the `:in => [true, false]`

File: app/views/cart/show.html.erb [hunk 5]
```diff
@@ -15,8 +16,14 @@
       <td>
         <p align = "center">
           <span>
-            <img align = "center" id="cart-image" src= <%= item.image %>>
-            <%=link_to item.name, "/items/#{item.id}" %>
+            <ul><img align = "center" id="cart-image" src= <%= item.image %> style= "max-width:200px;width:100%" ></ul>
```
Comment (LHJE) [inline app/views/cart/show.html.erb:19]: Great styling!

File: spec/features/cart/checkout_spec.rb [hunk 6]
```diff
@@ -1,39 +1,199 @@
...
       click_on "Add To Cart"
-      visit "/items/#{@pencil.id}"
-      click_on "Add To Cart"
+
+      3.times do
+        visit "/items/#{@pencil.id}"
+        click_on "Add To Cart"
+      end
+
       @items_in_cart = [@paper,@tire,@pencil]
```
Comment (LHJE) [inline spec/features/cart/checkout_spec.rb:35]: We may not need this anymore.

File: spec/features/cart/checkout_spec.rb [hunk 7]
```diff
@@ -1,39 +1,199 @@
...
 
-    it 'Theres a link to checkout' do
+    it "requires a visitor to login or register before checking out" do
       visit "/cart"
 
-      expect(page).to have_link("Checkout")
+      expect(page).to_not have_link("Checkout")
+      expect(page).to have_content("You must Register or Log In to finish checking out.")
+      expect(page).to have_link("Register")
+      expect(page).to have_link("Log In")
```
Comment (LHJE) [inline spec/features/cart/checkout_spec.rb:44]: Maybe be a little more specific about clicking the link in the `#checkout-authorization`

File: spec/features/cart/checkout_spec.rb [hunk 8]
```diff
@@ -1,39 +1,199 @@
...
+
+        within 'nav' do
+          click_link "Login"
+        end
+
+        fill_in :email, with: @user.email
+        fill_in  :password, with: @user.password
+        click_button "Login"
+
+        visit "/cart"
```
Comment (LHJE) [inline spec/features/cart/checkout_spec.rb:81]: Let's try to implement this to get rid of lines 71-79:
`allow_any_instance_of(ApplicationController).to receive(:current_user).and_return(@bike_shop_employee)`

File: spec/features/cart/checkout_spec.rb [hunk 9]
```diff
@@ -1,39 +1,199 @@
...
+        @order_2.item_orders.create!(item: @tire, price: @tire.price, quantity: 2)
+        @order_3.item_orders.create!(item: @pencil, price: @pencil.price, quantity: 5)
+
+        visit "/"
+        within 'nav' do
+          click_link "Login"
+        end
+        fill_in :email, with: @user.email
+        fill_in  :password, with: @user.password
+        click_button "Login"
```
Comment (LHJE) [inline spec/features/cart/checkout_spec.rb:178]: Ooo let's also replace this with the new log in fix.

**Review by priyapower — COMMENTED** [review]
Looks perfect - thanks for the code review

File: app/controllers/orders_controller.rb (hunk 1 above)
Comment (priyapower) [reply app/controllers/orders_controller.rb:15]: Updated by @LHJE

File: app/controllers/orders_controller.rb (hunk 2 above)
Comment (priyapower) [reply app/controllers/orders_controller.rb:22]: Updated by @LHJE - actually likes 22-23 in our atom

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (priyapower) [reply app/controllers/users_controller.rb:13]: Leaving this for refactor!

File: spec/features/cart/checkout_spec.rb (hunk 6 above)
Comment (priyapower) [reply spec/features/cart/checkout_spec.rb:35]: Updated by @LHJE

File: spec/features/cart/checkout_spec.rb (hunk 7 above)
Comment (priyapower) [reply spec/features/cart/checkout_spec.rb:44]: Updated nested expectations under within CSS by @LHJE

File: spec/features/cart/checkout_spec.rb (hunk 8 above)
Comment (priyapower) [reply spec/features/cart/checkout_spec.rb:81]: Updated by @LHJE

File: spec/features/cart/checkout_spec.rb (hunk 9 above)
Comment (priyapower) [reply spec/features/cart/checkout_spec.rb:178]: Updated by @LHJE

**Pull request merged**